lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/DefaultVortexMaster.java

```diff
@@ -140,4 +153,30 @@ public void taskletCancelled(final String workerId, final int taskletId) {
   public void terminate() {
     runningWorkers.terminate();
   }
+
+  private synchronized void putDelegate(final Tasklet tasklet, final VortexFutureDelegate delegate) {
+    taskletFutureMap.put(tasklet.getId(), delegate);
+  }
+
+  private synchronized VortexFutureDelegate fetchDelegate(final List<Integer> taskletIds) {
+    synchronized (taskletFutureMap) {
+      VortexFutureDelegate delegate = null;
+      for (final int taskletId : taskletIds) {
+        final VortexFutureDelegate currDelegate = taskletFutureMap.remove(taskletId);
+        if (currDelegate == null) {
+          // TODO[JIRA REEF-500]: Consider duplicate tasklets.
+          throw new RuntimeException("Tasklet should only be removed once.");
+        }
+
+        if (delegate == null) {
+          delegate = currDelegate;
+        } else {
+          assert delegate == currDelegate;
```

Comment:
For local aggregation, I'm considering using the continuation token pattern. I envision having a blocking `TOutput get(ContinuationToken)` call, which will block until a Worker reports back a list of Tasklets associated with a Future, We will record the state of the aggregation future in the `ContinuationToken`. This is such that we don't need to wait until all the Tasklets associated with the aggregation function return in order to produce a result for the user. Local aggregation will also support callbacks.